File: MorePopularGitHubApp/app/src/main/java/com/github/api/morepopulargithubapp/adapter/RepositoryAdapter.java
package com.github.api.morepopulargithubapp.adapter;

import android.content.Context;
import android.view.ViewGroup;

import com.github.api.morepopulargithubapp.model.vo.Repository;
import com.github.api.morepopulargithubapp.view.RepositoryItemView;
import com.github.api.morepopulargithubapp.view.RepositoryItemView_;

import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.RootContext;

import java.util.List;


@EBean
public class RepositoryAdapter extends RecyclerViewAdapterBase<Repository, RepositoryItemView> {

    @RootContext
    Context context;

    @Override
    protected RepositoryItemView onCreateItemView(ViewGroup parent, int viewType) {
        return RepositoryItemView_.build(context);
    }

    @Override
    public void onBindViewHolder(ViewWrapper<RepositoryItemView> holder, int position) {
          RepositoryItemView view = holder.getView();
          Repository repository = items.get(position);
          view.bind(repository);
    }

    @Override
    public void setItems(List<Repository> items) {
        super.setItems(items);
    }

    @Override
    public long getItemId(int position) {
        return super.getItemId(position);
    }

    @Override
    public int getItemViewType(int position) {
        return super.getItemViewType(position);
    }
}

File: app/src/main/java/com/joaoibarra/gitapp/features/repository/RepositoryAdapter.kt
package com.joaoibarra.gitapp.features.repository

import android.arch.paging.PagedListAdapter
import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.joaoibarra.gitapp.R
import com.joaoibarra.gitapp.api.model.Repo
import com.joaoibarra.gitapp.extensions.loadCircle

import kotlinx.android.synthetic.main.item_repository.view.*

class RepositoryAdapter(val listener: (Repo?) -> Unit) : PagedListAdapter<Repo, RepositoryAdapter.ItemViewHolder>(itemDiff) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_repository, parent, false)
        return ItemViewHolder(view)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        holder.bind(getItem(position), listener)
    }

    class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(repository: Repo?, listener: (Repo?) -> Unit) = with(itemView) {
            tvName.text = repository?.name
            tvDescription.text = repository?.description
            tvForkCount.text = repository?.forksCount
            tvStarGazerCount.text = repository?.stargazersCount
            tvOwnerLogin.text = repository?.owner?.login
            ivAvatar.loadCircle(repository?.owner?.avatarUrl)
            setOnClickListener { listener(repository) }
        }
    }

    companion object {
        val itemDiff = object: DiffUtil.ItemCallback<Repo>() {

            override fun areItemsTheSame(old: Repo, new: Repo): Boolean {
                return old.id == new.id
            }

            override fun areContentsTheSame(old: Repo, new: Repo): Boolean {
                return old == new
            }

        }
    }

}
